src/metabase/automagic_dashboards/filters.clj

```diff
@@ -175,28 +176,42 @@
-    (if (-> filter-clause first qp.util/normalize-token (= :and))
-      (mapcat flatten-filter-clause (rest filter-clause))
+  "Returns a sequence of filter subclauses making up `filter-clause` by flattening `:and` compound filters.
+
+    (flatten-filter-clause [:and
+                            [:= [:field-id 1] 2]
+                            [:and
+                             [:= [:field-id 3] 4]
+                             [:= [:field-id 5] 6]]])
+    ;; -> ([:= [:field-id 1] 2]
+           [:= [:field-id 3] 4]
+           [:= [:field-id 5] 6])"
+  [[clause-name, :as filter-clause]]
+  (when (seq filter-clause)
+    (if (= clause-name :and)
+      (rest (mbql.u/simplify-compound-filter filter-clause))
       [filter-clause])))
 
 (defn inject-refinement
-  "Inject a filter refinement into an MBQL filter clause.
-   There are two reasons why we want to do this: 1) to reduce visual noise when we display applied
-   filters; and 2) some DBs don't do this optimization or even protest (eg. GA) if there are
-   duplicate clauses.
-
-   Assumes that any refinement sub-clauses referencing fields that are also referenced in the
-   main clause are subsets of the latter. Therefore we can rewrite the combined clause to ommit
-   the more broad version from the main clause.
-   Assumes both filter clauses can be flattened by recursively merging `:and` claueses
-   (ie. no `:and`s inside `:or` or `:not`)."
+  "Inject a filter refinement into an MBQL filter clause, returning a new filter clause.
+
+  There are two reasons why we want to do this: 1) to reduce visual noise when we display applied filters; and 2) some
+  DBs don't do this optimization or even protest (eg. GA) if there are duplicate clauses.
+
+  Assumes that any refinement sub-clauses referencing fields that are also referenced in the main clause are subsets
+  of the latter. Therefore we can rewrite the combined clause to ommit the more broad version from the main clause.
+  Assumes both filter clauses can be flattened by recursively merging `:and` claueses
+  (ie. no `:and`s inside `:or` or `:not`)."
   [filter-clause refinement]
-  (let [in-refinement? (into #{}
-                         (map collect-field-references)
-                         (flatten-filter-clause refinement))]
-    (->> filter-clause
-         flatten-filter-clause
-         (remove (comp in-refinement? collect-field-references))
-         (apply merge-filter-clauses refinement))))
+  (let [in-refinement?  (into #{}
+                              (map collect-field-references)
+                              (flatten-filter-clause refinement))
+        existing-filters (->> filter-clause
+                             flatten-filter-clause
+                             (remove (comp in-refinement? collect-field-references)))]
+    (if (seq existing-filters)
+      ;; for some reason existing-filters and refinement are more often than not non-normalized for reasons I don't
+      ;; understand.
```

Comment:
I think it's because in a large part (if xraying anything aside from cards or ad-hoc questions) they are programatically generated and have at this point never passed through normalization.

Reply:
thanks. I will update the comment.

Reply:
If my understanding is correct and queries attached to cards get normalized on load, it would make sense to add normalization to `models/query/adhoc-query` as well.